File: tracers/java/src/Inspector.java
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonPrimitive;
import com.sun.jdi.*;
import com.sun.jdi.event.ExceptionEvent;
import com.sun.jdi.event.LocatableEvent;

import java.util.*;
import java.util.stream.Collectors;

/**
 * Inspect events and produces maps with their state data.
 */
class Inspector {
    private long orderedIdCount;
    private Map<Long, String> orderedIds;
    private Map<Long, String> previousOrderedIds;

    /**
     * Initialize the inspector and ordered id generators.
     */
    public Inspector() {
        orderedIdCount = 0;
        orderedIds = new HashMap<>();
        previousOrderedIds = new HashMap<>();
    }

    /**
     * Inspect the event to build a map with state data.
     *
     * @param event event where the state data will be extracted from
     * @return the processed event data
     * @throws IncompatibleThreadStateException
     */
    JsonObject inspect(LocatableEvent event) throws IncompatibleThreadStateException {
        previousOrderedIds = orderedIds;
        orderedIds = new HashMap<>();

        var snapshot = new JsonObject();
        snapshot.addProperty("info", !(event instanceof ExceptionEvent) ? "ok" : "warn");
        var frames = new ArrayList<>(event.thread().frames());
        Collections.reverse(frames);
        snapshot.add(
                "stack",
                frames.stream()
                        .map(StackFrame::location)
                        .map(l -> {
                            var scope = new JsonObject();
                            scope.addProperty("line", l.lineNumber() - 1);
                            scope.addProperty("name", l.method().name());
                            return scope;
                        })
                        .collect(
                                () -> new JsonArray(frames.size()),
                                JsonArray::add,
                                (scope0, scope1) -> {
                                    throw new RuntimeException("parallel stream not allowed");
                                }
                        )
        );

        snapshot.add("heap", new JsonObject());
        var threadReference = frames.get(0).thread();
        int[] scopeIndex = {0};
        frames.stream()
                .map(f -> {
                    try {
                        return Map.entry(f.visibleVariables(), f.getValues(f.visibleVariables()));
                    } catch (AbsentInformationException e) {
                        throw new RuntimeException(e);
                    }
                })
                // frames get invalid after invoking methods in objects
                // because of this, all variables have to be collected before any evaluation
                .collect(Collectors.toList())
                .forEach(e -> {
                    var localVariables = e.getKey();
                    var values = e.getValue();
                    var scope = snapshot.get("stack").getAsJsonArray().get(scopeIndex[0]++).getAsJsonObject();
                    scope.add(
                            "members",
                            localVariables.stream()
                                    .map(localVariable -> {
                                        var member = new JsonObject();
                                        member.addProperty("key", localVariable.name());
                                        member.add("value", inspectValue(snapshot, values.get(localVariable), threadReference));
                                        return member;
                                    })
                                    .collect(
                                            () -> new JsonArray(localVariables.size()),
                                            JsonArray::add,
                                            (memberA, memberB) -> {
                                                throw new RuntimeException("parallel stream not allowed");
                                            }
                                    )
                    );
                });
        return snapshot;
    }

    /**
     * Recursively inspect values of the heap.
     * Mutates the snapshot if it is an object.
     *
     * @param snapshot        snapshot to be filled with heap information
     * @param jdiValue        value to be processed
     * @param threadReference debugee jvm thread to call functions on it
     * @return the transformed value
     */
    private JsonElement inspectValue(JsonObject snapshot, Value jdiValue, ThreadReference threadReference) {
        JsonElement value;
        return jdiValue == null
                ? new JsonPrimitive("null")
                : (value = inspectPrimitive(jdiValue)) != null
                ? value
                : (value = inspectObject(snapshot, jdiValue, threadReference)) != null
                ? value
                : new JsonPrimitive("unknown");
    }

    private JsonElement inspectPrimitive(Value jdiValue) {
        return jdiValue instanceof PrimitiveValue ?
                jdiValue instanceof BooleanValue
                        ? new JsonPrimitive(Boolean.toString(((BooleanValue) jdiValue).value()))
                        : jdiValue instanceof CharValue
                        ? new JsonPrimitive(Character.toString(((CharValue) jdiValue).value()))
                        : jdiValue instanceof ByteValue
                        ? new JsonPrimitive(((ByteValue) jdiValue).value())
                        : jdiValue instanceof ShortValue
                        ? new JsonPrimitive(((ShortValue) jdiValue).value())
                        : jdiValue instanceof IntegerValue
                        ? new JsonPrimitive(((IntegerValue) jdiValue).value())
                        : jdiValue instanceof FloatValue
                        ? new JsonPrimitive(((FloatValue) jdiValue).value())
                        : jdiValue instanceof DoubleValue
                        ? new JsonPrimitive(((DoubleValue) jdiValue).value())
                        : jdiValue instanceof LongValue
                        ? Math.abs(((LongValue) jdiValue).value()) < Math.pow(2, 53)
                        ? new JsonPrimitive(((LongValue) jdiValue).value())
                        : new JsonPrimitive(Long.toString(((LongValue) jdiValue).value()))
                        : null
                : null;
    }

    private JsonElement inspectObject(JsonObject snapshot, Value jdiValue, ThreadReference threadReference) {
        JsonElement value;
        if (!(jdiValue instanceof ObjectReference)) return null;
        var jdiObjRef = (ObjectReference) jdiValue;
        var stringObj = inspectString(jdiObjRef);
        if (stringObj != null) return stringObj;
        var boxedObj = inspectBoxed(jdiObjRef);
        if (boxedObj != null) return boxedObj;

        var id = jdiObjRef.uniqueID();
        var orderedId = "";
        if (orderedIds.containsKey(id))
            orderedId = orderedIds.get(id);
        else if (previousOrderedIds.containsKey(id)) {
            orderedIds.put(id, previousOrderedIds.get(id));
            orderedId = orderedIds.get(id);
        } else {
            orderedIds.put(id, Long.toString(orderedIdCount++));
            orderedId = orderedIds.get(id);
        }

        if (snapshot.get("heap").getAsJsonObject().has(orderedId)) {
            var idValue = new JsonArray(1);
            idValue.add(id);
            return idValue;
        }

        return (value = inspectArray(snapshot, jdiObjRef, threadReference)) != null
                ? value
                : (value = inspectCollection(snapshot, jdiObjRef, threadReference)) != null
                ? value
                : (value = inspectMap(snapshot, jdiObjRef, threadReference)) != null
                ? value
                : (value = inspectUserObject(snapshot, jdiObjRef, threadReference)) != null
                ? value
                : new JsonPrimitive("class " + jdiObjRef.referenceType().name());
    }

    private JsonElement inspectString(ObjectReference jdiObjRef) {
        return jdiObjRef instanceof StringReference ? new JsonPrimitive(((StringReference) jdiObjRef).value()) : null;
    }

    private JsonElement inspectBoxed(ObjectReference jdiObjRef) {
        try {
            var jdiObjRefClass = Class.forName(jdiObjRef.referenceType().name());
            return Boolean.class.isAssignableFrom(jdiObjRefClass) || Character.class.isAssignableFrom(jdiObjRefClass) ||
                    Byte.class.isAssignableFrom(jdiObjRefClass) || Short.class.isAssignableFrom(jdiObjRefClass) ||
                    Integer.class.isAssignableFrom(jdiObjRefClass) || Long.class.isAssignableFrom(jdiObjRefClass) ||
                    Float.class.isAssignableFrom(jdiObjRefClass) || Double.class.isAssignableFrom(jdiObjRefClass)
                    ? inspectPrimitive(jdiObjRef.getValue(jdiObjRef.referenceType().fieldByName("value")))
                    : null;
        } catch (ClassNotFoundException e) {
            // array types always throw this exception
        }
        return null;
    }

    private JsonElement inspectIterable(JsonObject snapshot, ArrayReference jdiArrRef, String id, String type, String category, ThreadReference threadReference) {
        var obj = new JsonObject();
        // add id to heap graph (it has to be added before other objects inspections)
        snapshot.get("heap").getAsJsonObject().add(id, obj);
        obj.addProperty("type", type);
        obj.addProperty("category", category);
        int[] memberIndex = {0};
        obj.add(
                "members",
                jdiArrRef.getValues().stream()
                        .map(jdiValue -> {
                            var member = new JsonObject();
                            member.addProperty("key", memberIndex[0]++);
                            member.add("value", inspectValue(snapshot, jdiValue, threadReference));
                            return member;
                        })
                        .collect(
                                () -> new JsonArray(jdiArrRef.length()),
                                JsonArray::add,
                                (member0, member1) -> {
                                    throw new RuntimeException("parallel stream not allowed");
                                }
                        )
        );
        var idValue = new JsonArray(1);
        idValue.add(id);
        return idValue;
    }

    private JsonElement inspectArray(JsonObject snapshot, ObjectReference jdiObjRef, ThreadReference threadReference) {
        return jdiObjRef instanceof ArrayReference
                ? inspectIterable(
                snapshot,
                (ArrayReference) jdiObjRef,
                orderedIds.get(jdiObjRef.uniqueID()),
                jdiObjRef.referenceType().name(),
                "list",
                threadReference
        )
                : null;
    }

    private JsonElement inspectCollection(JsonObject snapshot, ObjectReference jdiObjRef, ThreadReference threadReference) {
        try {
            var jdiObjRefClass = Class.forName(jdiObjRef.referenceType().name());
            return Collection.class.isAssignableFrom(jdiObjRefClass)
                    ? inspectIterable(
                    snapshot,
                    (ArrayReference) jdiObjRef.invokeMethod(
                            threadReference,
                            jdiObjRef.referenceType().methodsByName("toArray", "()[Ljava/lang/Object;").get(0),
                            List.of(),
                            ObjectReference.INVOKE_SINGLE_THREADED
                    ),
                    orderedIds.get(jdiObjRef.uniqueID()),
                    jdiObjRef.referenceType().name(),
                    List.class.isAssignableFrom(jdiObjRefClass) ? "list" : "set",
                    threadReference
            )
                    : null;
        } catch (ClassNotFoundException | InvalidTypeException | ClassNotLoadedException | IncompatibleThreadStateException | InvocationException e) {
            // ignore errors
        }
        return null;
    }

    private JsonElement inspectMap(JsonObject snapshot, ObjectReference jdiObjRef, ThreadReference threadReference) {
        try {
            var jdiObjRefClass = Class.forName(jdiObjRef.referenceType().name());
            if (!Map.class.isAssignableFrom(jdiObjRefClass)) return null;

            var jdiSetRef = (ObjectReference) jdiObjRef.invokeMethod(
                    threadReference,
                    jdiObjRef.referenceType().methodsByName("entrySet").get(0),
                    List.of(),
                    ObjectReference.INVOKE_SINGLE_THREADED
            );
            var jdiArrRef = (ArrayReference) jdiSetRef.invokeMethod(
                    threadReference,
                    jdiSetRef.referenceType().methodsByName("toArray", "()[Ljava/lang/Object;").get(0),
                    List.of(),
                    ObjectReference.INVOKE_SINGLE_THREADED
            );
            var id = orderedIds.get(jdiObjRef.uniqueID());
            var obj = new JsonObject();
            // add id to heap graph (it has to be added before other objects inspections)
            snapshot.get("heap").getAsJsonObject().add(id, obj);
            obj.addProperty("type", jdiObjRef.referenceType().name());
            obj.addProperty("category", "map");
            obj.add(
                    "members",
                    jdiArrRef.getValues().stream()
                            .map(jdiValue -> {
                                try {
                                    var jdiValueEntry = (ObjectReference) jdiValue;
                                    var jdiValueEntryKey = jdiValueEntry.invokeMethod(
                                            threadReference,
                                            jdiValueEntry.referenceType().methodsByName("getKey").get(0),
                                            List.of(),
                                            ObjectReference.INVOKE_SINGLE_THREADED
                                    );
                                    var jdiValueEntryValue = jdiValueEntry.invokeMethod(
                                            threadReference,
                                            jdiValueEntry.referenceType().methodsByName("getValue").get(0),
                                            List.of(),
                                            ObjectReference.INVOKE_SINGLE_THREADED
                                    );
                                    var member = new JsonObject();
                                    member.add("key", inspectValue(snapshot, jdiValueEntryKey, threadReference));
                                    member.add("value", inspectValue(snapshot, jdiValueEntryValue, threadReference));
                                    return member;
                                } catch (InvalidTypeException | ClassNotLoadedException | IncompatibleThreadStateException | InvocationException e) {
                                    throw new RuntimeException(e);
                                }
                            })
                            .collect(
                                    () -> new JsonArray(jdiArrRef.length()),
                                    JsonArray::add,
                                    (member0, member1) -> {
                                        throw new RuntimeException("parallel stream not allowed");
                                    }
                            )
            );
            var idValue = new JsonArray(1);
            idValue.add(id);
            return idValue;
        } catch (ClassNotFoundException | InvalidTypeException | ClassNotLoadedException | IncompatibleThreadStateException | InvocationException e) {
            // ignore errors
        }
        return null;
    }

    private JsonElement inspectUserObject(JsonObject snapshot, ObjectReference jdiObjRef, ThreadReference threadReference) {
        var jdiObjClassRefData = jdiObjRef.referenceType().toString().split(" ");

        // check if is a class and is defined in the default package (user defined)
        if (!jdiObjClassRefData[0].equals("class") || jdiObjClassRefData[1].contains(".")) return null;

        var orderedFields = jdiObjRef.referenceType().allFields();
        var fieldsValues = jdiObjRef.getValues(orderedFields);
        var id = orderedIds.get(jdiObjRef.uniqueID());
        var obj = new JsonObject();
        // add id to heap graph (it has to be added before other objects inspections)
        snapshot.get("heap").getAsJsonObject().add(id, obj);
        obj.addProperty("type", jdiObjRef.referenceType().name());
        obj.addProperty("category", "map");
        obj.add(
                "members",
                orderedFields.stream()
                        .map(field -> {
                            var member = new JsonObject();
                            member.addProperty("key", field.name());
                            member.add("value", inspectValue(snapshot, fieldsValues.get(field), threadReference));
                            return member;
                        })
                        .collect(
                                () -> new JsonArray(orderedFields.size()),
                                JsonArray::add,
                                (member0, member1) -> {
                                    throw new RuntimeException("parallel stream not allowed");
                                }
                        )
        );
        var idValue = new JsonArray(1);
        idValue.add(id);
        return idValue;
    }
}

File: tracers/python/src/inspector.py
import sys
import types


class Inspector:
    """
    Inspect frames and produces snapshots with the state.
    """

    def __init__(self):
        """
        Initialize the inspector and the ordered id generators.
        """
        self.ordered_id_count = 0
        self.ordered_ids = {}
        self.previous_ordered_ids = {}

    # types that can be converted to a JSON string
    STRING_LIKE_TYPES = (bool, complex, str, type(None))

    # types that can be converted to a JSON number (within javascript's restriction of +-2^53 - 1)
    NUMBER_LIKE_TYPES = (int, float)

    def inspect(self, frame: types.FrameType, event: str, args, exec_frame: types.FrameType):
        """
        Inspect the frame and its event to build a snapshot with state.
        - frame: `FrameType`: frame where the state data will be extracted from
        - event: `str`: frame event type, one of: call, line, exception or return
        - args: `any | any[t]`: return object if event is return or exception data if event is exception
        - exec_frame: `FrameType`: frame of the exec() call, it is used to know which frame is the base frame
        - return `dict`: the snapshot
        """
        self.previous_ordered_ids = self.ordered_ids
        self.ordered_ids = {}

        frames = self._collect_frames(frame, exec_frame)
        stack = self._create_stack(frames)
        heap = self._create_heap(stack, frames, event, args)
        return {'event': event, 'stack': stack, 'heap': heap}

    def _collect_frames(self, frame: types.FrameType, stop_frame: types.FrameType):
        """
        Collect all frames of the same file until a stop frame is found.
        - frame: `FrameType`: frame to start collecting parents
        - stop_frame: `FrameType`: stops the collecting frame if stop_frame is found
        - return `FrameType[]`: collected frames
        """
        filename = frame.f_code.co_filename
        frames = []
        while frame is not None and frame != stop_frame and frame.f_code.co_filename == filename:
            frames.append(frame)
            frame = frame.f_back
        frames.reverse()
        return frames

    def _create_stack(self, frames: list):
        """
        Creates a list of dicts with the names and first lines of the the received frames
        - frames `FrameType[]` collected frames
        - return `dict`: collected stack data
        """
        return [{'line': frame.f_lineno - 1, 'name': frame.f_code.co_name} for frame in frames]

    def _create_heap(self, stack: list, frames: list, event: str, args):
        """
        Create and populate the heap, and also set the members of stack scopes.
        - stack `dict`: collected stack data from frames
        - frames `FrameType[]` collected frames
        - event: `str`: frame event type, one of: call, line, exception or return
        - args: `any | any[t]`: return object if event is return or exception data if event is exception
        - return `dict`: heap containing objects
        """
        heap = {}
        module = frames[-1].f_globals['__name__']
        classes = set()
        for scope, frame in zip(stack, frames):
            variables = frame.f_locals
            scope['members'] = [
                {'key': key, 'value': self._inspect_object(heap, variables[key], classes, module)}
                for key, value in variables.items() if not key.startswith('_')
            ]
        if event == 'return' and len(stack) > 1:
            stack[-1]['members'].append({'key': '#return#', 'value': self._inspect_object(heap, args, classes, module)})
        return heap

    def _inspect_object(self, heap: dict, obj, classes: set, module: str):
        """
        Recursively inspect objects in the heap.
        String and Number like objects are transformed in str or int.
        Complex objects are transformed in dictionaries that contain information of their type and members and added to
        the snapshot, then their references are returned as a list.
        - heap: `dict`: heap dict to be filled with obj information
        - obj: `*any`: object to be processed
        - classes: `set<str>`: set of user defined classes, which is used to known which objects shall be analyzed
        - module: main module name, used to only save classes that where declared in it
        - return `int | float | str | [str]`: the transformed value
        """
        if isinstance(obj, Inspector.STRING_LIKE_TYPES):
            return str(obj)
        elif isinstance(obj, Inspector.NUMBER_LIKE_TYPES):
            return obj if abs(obj) < 2 ** 53 else str(obj)
        elif isinstance(obj, type):
            if obj.__module__ == module:
                classes.add(obj)
            return str(obj)

        id_ = str(id(obj))
        if id_ in self.ordered_ids:
            ordered_id = self.ordered_ids[id_]
        elif id_ in self.previous_ordered_ids:
            ordered_id = self.ordered_ids[id_] = self.previous_ordered_ids[id_]
        else:
            ordered_id = self.ordered_ids[id_] = self.ordered_id_count
            self.ordered_id_count += 1

        if ordered_id in heap:
            return [ordered_id]

        type_ = type(obj).__name__
        category = 'other'
        members = None

        if isinstance(obj, (tuple, list, set)):
            category = 'list' if not isinstance(obj, set) else 'set'
            members = [*enumerate(obj)]
        elif isinstance(obj, dict):
            category = 'map'
            members = [*obj.items()]
        elif isinstance(obj, (*classes,)):
            category = 'map'
            members = [(key, value) for key, value in vars(obj).items() if not key.startswith('_')]

        if members is not None:  # known object type
            # add object id to the heap before further inspections to avoid stack overflows
            obj = heap[ordered_id] = {}
            obj['id'] = ordered_id
            obj['category'] = category
            obj['type'] = type_
            obj['members'] = [
                {
                    'key': self._inspect_object(heap, key, classes, module),
                    'value': self._inspect_object(heap, value, classes, module)
                }
                for key, value in members
            ]
            return [ordered_id]
        else:  # unknown object type
            # instead of inspecting unknown objects, inspect their type only
            return self._inspect_object(heap, type(obj), classes, module)
